rfc/resolve-directive.md

```diff
@@ -0,0 +1,189 @@
+
+Breaking change detection can use schema usage from these "internal" calls to prevent production issues.
+Query plans can show these implicit dependencies.
+A dashboard can show which queries depend on which services, explicit and implicit.
+Another dashboard can show dependencies between services.
+
+## Example 1: Validate a Reference
+
+```graphql
+# Review Subgraph
+type Review @key(fields: "id") {
+    id: ID!
+    rating: Int!
+    comment: String!
+}
+
+extend type Product @key(fields: "id") {
+    id: ID!
+    reviews: [Review]
+}
+
+type Query {
+    reviewByID(id: ID!): Review @inaccessible
+}
+```
+
+```graphql
+# Product Admin Subgraph
+
+input RemoveReviewInput {
+    productID: ID!
+    reviewID: ID!
+    # This field is used to validate the review exists
+    # if the Router resolves it to null, the mutation will fail because the field is not nullable
+    # if it's resolved to a value, the mutation will have access to the content as part of the input
+    review: RemoveReviewReviewInput! @inaccessible @resolve(query: "query Review($id: ID!) { reviewByID(id: $id) { id }}", variables: "{ \"id\": {{ .entity.reviewID }} }")
+}
+
+input RemoveReviewReviewInput {
+    id: ID!
+}
+
+type Mutation {
+    removeReview(input: RemoveReviewInput!): Review
+}
+```
+
+## Example 2: Query input from another Subgraph
+
+Let's say we've got a userInfo Subgraph which is able to provide user information based on request headers.
+We'd like to make this information available to all other Subgraphs without exposing it to the public API.
+
+```graphql
+# UserInfo Subgraph
+type Query {
+    userInfo: UserInfo @inaccessible
+}
+
+type UserInfo @inaccessible {
+    id: ID!
+    name: String!
+    email: String!
+}
+```
+
+We've defined a UserInfo Subgraph which provides user information.
+We're market the `userInfo` field and the `UserInfo` type as `@inaccessible` to prevent clients from accessing this information directly.
+Let's see how we can use it in another Subgraph:
+
+```graphql
+# User Subgraph
+type Query {
+    currentUser(input: CurrentUserInput! @inaccessible): User
+}
+
+input CurrentUserInput {
+    info: UserInfoInput @inaccessible @resolve(query: "query { info: userInfo { id name email }}")
+}
+
+input UserInfoInput {
+    id: ID!
+    name: String!
+    email: String!
+}
+
+type User {
+    id: ID!
+    name: String!
+    email: String!
+}
+```
+
+The public schema of the Supergraph will look as follows:
+
+```graphql
+type Query {
+    currentUser: User
+}
+
+type User {
+    id: ID!
+    name: String!
+    email: String!
+}
+```
+
+As an alternative, it's also possible to use the `@resolve` directive on a field to resolve the value of the field.
+
+```graphql
+# User Subgraph
+type User @key(fields: "id") {
+    id: ID!
+    name: String! @requires(fields: "info")
+    email: String! @requires(fields: "info")
+    info: UserInfo @inaccessible @resolve(query: "query { info: userInfo { id name email }}")
```

Comment:
So here, `User.info` is delegating to `@resolve`, which is calling `Query.userInfo`? Although I don't see any queries defined in this subgraph—does that mean the query can be defined anywhere?

Reply:
And if you can use any query defined anywhere, how do you make sure the mapping/relationship is correct?

Reply:
Yes, you can query the whole graph.
Composition needs to validate this query.